### DISCUSSION TITLE
자바스크립트에서 배열을 순회하는 함수를 설명해주세요

### DISCUSSION THREAD

**Discussion by skyhanull** [wanted-9th-3team/frontend-study#22] [프론트 기술 질문 공간]
자바스크립트에서 배열을 순회하는 함수를 설명해주세요
그리고 그 중 map과 foreach의 차이점도 포함해서 작성해주세요

**Comment by skyhanull** [comment]
자바스크립트에서 배열을 순회하는 메소드에는 filter,map,foreach,reduce,,등 이 존재합니다
그 중에서 map,foreach는 둘다 배열을 순회하는 메소드이지만 다른 점은 
forEach는 콜백함수를 인자로 받고 각 요소에 콜백함수를 실행시킨다. return값이 undefined로 나온다
map은 콜백함수를 인자로 받고 각 요소에 콜백함수를 실행시킨 새로운 배열을 반환한다는 차이가 있습니다

**Comment by michoball** [comment]
배열을 순회하는 것만 생각한다면 filter,  find,  map,  reduce,  some,  every, foreach 의 함수가 있겠습니다. 각각 세부적인 특징은 다르지만 배열을 돌면서 특정 작업을 수행한 결과를 도출할 때 사용합니다. 

그중 foreach와 map은 return 하는 값이 있냐 없냐의 차이가 있겠습니다.  foreach는 배열을 순회하며 콜백함수를 실행하기만 하고, map은 이 콜백함수를 실행한 결과를 가지는 새로운 배열을 리턴합니다.

**Comment by WooGie911** [comment]
자바스크립트에서 배열을 순회하는 함수는 워낙 많겠지만 제가 아는 함수만 말해보자면
배열 내의 모든 요소 각각에 대하여 
조건을 만족하는 새로운 배열을 반환하는 map, 
조건에 해당하는 배열을 반환하는 filter, 
조건을 만족하는 첫번쨰 요소의 값을 반환하는 find, 
조건을 만족하는 첫번쨰 요소의 인덱스를 반환하는 findIndex 등이 있습니다.

forEach와 map의 차이점은 map은 새로운 배열을 반환하는 반면 forEach는 새로운 배열을 반환하는 것이 아닌 배열 내의 모든 요소를 순회하는 for문 이라고 볼 수 있습니다.

**Comment by BaekYuri** [comment]
배열을 순회하는 함수로 filter, find, findIndex, forEach, map 등이 있습니다.
조건에 맞는 배열 값만 걸러내서 return 하는 filter 함수,
조건에 맞는 첫번째 요소의 값만 return하는 find 함수, 그 index를 return하는 findIndex 함수,
배열을 순회하면서 원하는 일을 수행할 수 있는 forEach, 순회하면서 새로운 데이터의 배열을 return하는 map 함수가 있습니다.
forEach는 return값이 없는 반면에, map은 새로운 배열을 반환합니다.

**Comment by bona373737** [comment]
foreach : 원본배열이 수정된다. 
map,filter : 원본배열의 변경없이, 배열 순회 결과값을 새로운 배열로 리턴 받는다.
